core/src/Kokkos_Core.hpp

```diff
@@ -99,6 +99,9 @@ void declare_configuration_metadata(const std::string& category,
 [[nodiscard]] bool is_initialized() noexcept;
 [[nodiscard]] bool is_finalized() noexcept;
 
+[[nodiscard]] int device_id() noexcept;
```

Comment:
I was thinking yesterday, if it would make sense to wrap the device id into a class so that one could have a different integer identifier. For example one could have that as far as kokkos is concerned, we have device 0,1,2... but the enumeration is different natively somehow. 
Something like:

```cpp
struct DeviceIdWrapper{
  int kokkos_based_id;
  int native_device_id; 
};

DeviceIdWrapper device_id() noexcept;
```

I don't know if this can be useful always, but it would help us guard against possible differences

Reply:
As you pointed out the native id will depend on the backend, and IMO it belongs to the interoperability facility that the backend is allowed to provide and by definition is not portable.
In the snippet you posted the `native_device_id` would need to be an execution space specific type instead of an `int`.  It might be my lack of imagination but I fail to see how one would do something useful with this information.

Reply:
Having the native thing available facilitates interoperability with non-kokkos code that wants to make sure it's using the right resources (either the same ones, or distinct ones)

Reply:
We agreed that a native thing would be useful, but is not necessary to address the current need.